server/server.go

```diff
@@ -1918,6 +1918,12 @@ func (s *Server) createClient(conn net.Conn, ws *websocket) *client {
 	s.totalClients++
 	s.mu.Unlock()
 
+	// Attempt to authorize with cookie JWT
```

Comment:
I have a question on what should be the sever behavior: if user configures `jwt_cookie` but trusted keys are not configured, should we fail the config (say in validateWebsocketOptions())?

I am concerned of calling isClientAuthorized() which would end-up registering a user so early in the process. I would rather wait and do this original place, which is when receiving the CONNECT protocol.

If we were to do that, that means removing the code here and "move" it to processConnect() before releasing the client's lock:
```
...
accountNew := c.opts.AccountNew
// For headers both client and server need to support.
c.headers = supportsHeaders && c.opts.Headers
// For websocket clients, if there is a cookieJwt, authenticate with that.
if c.ws != nil && c.ws.cookieJwt != "" {
    c.opts.JWT = c.ws.cookieJwt
}
ujwt := c.opts.JWT
c.mu.Unlock()
```
Now, it is understood that it means no fallback to other authentication mode, which I honestly think that it is right because if a client provides a jwt cookie, then we should use that and only that. (also it means that we would need to update your tests to remove the "fallback" situations).

Question also is above, if there was a JWT present in the CONNECT, should we use that one or the one from c.ws.cookieJWT? In other words, which one should override?

What do you think? Also pinging Derek here: @derekcollison

Reply:
- Config with jwt_cookie and no trusted keys should be invalid, agreed.
- `CONNECT {}\r\n` is an auth flow that wasn't used before, as far as I know. But other than that, it's noninvasive enough, ad-hoc as it is.
- Lack of graceful auth fallback will make the code more opaque if fallback necessary, making client code require a much more expensive re-connect after `CONNECT {}\r\n -> -ERR`. The auth_required field in INFO also loses its purpose.
- JWT in CONNECT should have priority. HttpOnly cookies can't be removed or modified client-side, and if JWT is in CONNECT, the client knows what it's doing.

I haven't found downsides to early user registration. My understanding is,
- When server is accepting connections, all user configuration should already be loaded and initialized.
- createClient() [the parent method] is called in a separate goroutine, so it won't lock up the accept loop.
- isClientAuthorized is called very close to where CONNECT could be handled (after TLS handshake, inside go readLoop()->c.parse)

I'll implement everything once we have a consensus on CONNECT usage.

Reply:
> CONNECT {}\r\n is an auth flow that wasn't used before, as far as I know. But other than that, it's noninvasive enough, ad-hoc as it is.

Not sure what you mean here. CONNECT is expected from all our client libraries. They carry also information about what a client can do (say support for this or that feature).

> Lack of graceful auth fallback will make the code more opaque if fallback necessary, making client code require a much more expensive re-connect after CONNECT {}\r\n -> -ERR. 

My point was that I don't think we should be doing fallback. That is, it becomes difficult to know if a client was authenticated because it had a valid cookie jwt or because it fell back to connect auth which was accepted, say because websocket's auth did not have auth override, etc..

> The auth_required field in INFO also loses its purpose.

Understood. And although I don't think none of the client library we support are actually changing the way they sent the CONNECT based on that value, I get the point. But again, at this point in createClient() we know if the http request had a JWT cookie set, so we could say that if it is set in HTTP header's cookie, then this is what we are going to use to authenticate and set info.AuthRequired to false here, but then would have to take that into consideration in isClientAuthorized().

> When server is accepting connections, all user configuration should already be loaded and initialized.

Well yes, but this connection itself is far from being initialized. For instance c.initClient() has not even been invoked yet.

> createClient() [the parent method] is called in a separate goroutine, so it won't lock up the accept loop.

True, that is not a concern.

> isClientAuthorized is called very close to where CONNECT could be handled (after TLS handshake, inside go readLoop()->c.parse)

You mean pre-PR? Pre-PR it is invoked in processing of the CONNECT. Moving it in go readLoop()->c.parse is basically where it is now, when processing the CONNECT from the client.

Reply:
> Not sure what you mean here. CONNECT is expected from all our client libraries.

> And although I don't think none of the client library we support are actually changing the way they sent the CONNECT based on that value, I get the point

Ok, I didn't understand valid states for CONNECT properly. If every client always sends CONNECT, sending literally `CONNECT {}\r\n` (with empty parameters) shouldn't be an issue and is already used.

> My point was that I don't think we should be doing fallback.

Don't have a strong stance either way. Can go with no-fallback.

> c.initClient() has not even been invoked yet.

Yep, that's certainly concerning. A huge point to actually doing CONNECT. The fact it's not broken now doesn't mean that it'll remain this way.

> You mean pre-PR?

I meant in PR (inside createClient) relative to trunk (inside createClient, initClient, TLS, go readLoop()->c.parse). I thought it was close enough.

From the newly acquired info I stand on weak "auth on CONNECT" that you proposed for ease of initClient() maintainability.

Reply:
I am glad that you agree. I think that it will be safer (especially in the long term) if we keep the authentication/registration code where it has always been.
If there is trusted keys, then info.AuthRequired (in entering createClient() would be true). Regardless of that, we could make sure that if the ws.cookieJwt is not empty, it means that the client sent a JWT in the http request and we will expect a connect, even if in that case the connect will not really cary the auth's JWT.
Now if we get a push back from users that want to implement their own WS client library that don't want to send a CONNECT because the auth is in the http header, then we would need to authenticate toward the end of the createClient() function I guess.
But like I said, we are more and more relying on CONNECT to provide information on what the client can do feature-wise, so I think it is realistic to expect that "complex" clients (as opposed to say telnet) are going to send a CONNECT protocol.